x/withdraw/keeper/keeper.go

```diff
@@ -0,0 +1,123 @@
+	channelKeeper  channelkeeper.Keeper  // TODO: use interface
+	transferKeeper transferkeeper.Keeper // TODO: use interface
+	claimsKeeper   types.ClaimsKeeper
+}
+
+// NewKeeper returns keeper
+func NewKeeper(
+	ps paramtypes.Subspace,
+	ak types.AccountKeeper,
+	bk types.BankKeeper,
+	ics4Wrapper transfertypes.ICS4Wrapper,
+	ck channelkeeper.Keeper, // TODO: use interface
+	tk transferkeeper.Keeper, // TODO: use interface
+	claimsKeeper types.ClaimsKeeper,
+) *Keeper {
+	// set KeyTable if it has not already been set
+	if !ps.HasKeyTable() {
+		ps = ps.WithKeyTable(types.ParamKeyTable())
+	}
+
+	return &Keeper{
+		paramstore:     ps,
+		accountKeeper:  ak,
+		bankKeeper:     bk,
+		ics4Wrapper:    ics4Wrapper,
+		channelKeeper:  ck,
+		transferKeeper: tk,
+		claimsKeeper:   claimsKeeper,
+	}
+}
+
+// Logger returns logger
+func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
+}
+
+// IBC callbacks and transfer handlers
+
+// SendPacket implements the ICS4Wrapper interface from the transfer module.
+// It calls the underlying SendPacket function directly to move down the middleware stack.
+func (k Keeper) SendPacket(ctx sdk.Context, channelCap *capabilitytypes.Capability, packet exported.PacketI) error {
+	return k.ics4Wrapper.SendPacket(ctx, channelCap, packet)
+}
+
+// GetIBCDenomSource returns the source port and channel of the IBC denomination.
+// It returns an error if the the denomination is invalid or if the denom trace or source channel
+// is not found on the store.
+func (k Keeper) GetIBCDenomSource(ctx sdk.Context, denom, sender string) (srcPort, srcChannel string, err error) {
+	ibcDenom := strings.SplitN(denom, "/", 2)
+	if len(ibcDenom) < 2 {
+		return "", "", sdkerrors.Wrap(transfertypes.ErrInvalidDenomForTransfer, denom)
+	}
+
+	hash, err := transfertypes.ParseHexHash(ibcDenom[1])
+	if err != nil {
+		return "", "", sdkerrors.Wrapf(
+			err,
+			"failed to withdraw IBC vouchers back to sender '%s' in the corresponding IBC chain", sender,
+		)
+	}
+
+	denomTrace, found := k.transferKeeper.GetDenomTrace(ctx, hash)
+	if !found {
+		return "", "", sdkerrors.Wrapf(
+			transfertypes.ErrTraceNotFound,
+			"failed to withdraw IBC vouchers back to sender '%s' in the corresponding IBC chain", sender,
+		)
+	}
+
+	path := strings.Split(denomTrace.Path, "/")
+	if len(path)%2 != 0 {
+		return "", "", sdkerrors.Wrapf(
+			transfertypes.ErrInvalidDenomForTransfer,
+			"invalid denom (%s) trace path %s", denomTrace.BaseDenom, denomTrace.Path,
+		)
+	}
+
+	counterpartyPortID := path[0]
+	counterpartyChannelID := path[1]
+
+	channel, found := k.channelKeeper.GetChannel(ctx, counterpartyPortID, counterpartyChannelID)
+	if !found {
+		return "", "", sdkerrors.Wrapf(
+			channeltypes.ErrChannelNotFound,
+			"port ID %s, channel ID %s", counterpartyPortID, counterpartyChannelID,
+		)
+	}
+
+	srcPort = channel.Counterparty.PortId
+	srcChannel = channel.Counterparty.ChannelId
```

Comment:
Might be good to check if `channel.Counterparty.ChannelId` != nil. Optimistic handshakes could cause unforseen issues, but I'm pretty sure transfer doesn't support optimistic sends rn on ChanOpenInit